### PR TITLE
REGR: groupby doesn't identify null values when sort=False

### LINKED ISSUE #48506: BUG: groupby doesn't identify null values when sort=False
Closely related to #48476

As far as I can tell this only occurs when the input dtype to groupby is object.

```
df = pd.DataFrame({'a': [np.nan, pd.NA, None], 'b': [1, 2, 3]})
gb = df.groupby('a', sort=True, dropna=False)
print(gb.sum())

#      b
# a     
# NaN  6
```

but with `sort=False`:

```
df = pd.DataFrame({'a': [np.nan, pd.NA, None], 'b': [1, 2, 3]})
gb = df.groupby('a', sort=False, dropna=False)
print(gb.sum())

#       b
# a      
# NaN   1
# <NA>  2
# None  3
```

I think we should prefer the `sort=True` behavior as that is the default value for now, but prefer `sort=False` in the long run.

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #48506 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

No whatsnew since this regression was introduced in 1.5.0. This same change will also allow us to go forward with the cleanup in #48477 as it will no longer change behavior. I've also made sure the added test passes on 1.4.x.

### CODE REVIEW

File: pandas/tests/groupby/test_groupby_dropna.py [hunk 1]
```diff
@@ -454,3 +456,22 @@ def test_no_sort_keep_na(values, dtype, test_series):
...
+
+
+@pytest.mark.parametrize("test_series", [True, False])
+@pytest.mark.parametrize("dtype", [object, None])
+def test_null_is_null_for_dtype(
+    sort, dtype, unique_nulls_fixture, unique_nulls_fixture2, test_series
+):
+    # GH#48506 - groups should always result in using the null for the dtype
+    df = pd.DataFrame({"a": [1, 2]})
+    groups = pd.Series([unique_nulls_fixture, unique_nulls_fixture2], dtype=dtype)
```
Comment (mroeschke) [inline pandas/tests/groupby/test_groupby_dropna.py:468]: Would it be good to include `pd.NA` here (which isn't in unique_nulls_fixture)?

Comment (rhshadrach) [reply pandas/tests/groupby/test_groupby_dropna.py]: Absolutely - thanks, I didn't realize.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @rhshadrach

**Comment by phofl** [comment]
cc @rhshadrach This caused a couple of regressions, e.g. https://asv-runner.github.io/asv-collection/pandas/#algorithms.Hashing.time_series_string

Not sure if this was slower before, because history is missing a bit obviously

**Comment by rhshadrach** [comment]
Thanks @phofl - on my machine I'm seeing 2.3ms for the line `null_mask = isna(values)` whereas the whole added code is 2.58ms. My machine is likely a bit faster than the benchmark machine, so I believe this line is where most of the regression occurred. That many of the regressions don't involve null values confirms this.

I don't see a way around this op in general nor to speed it up - we could try to funnel the null values directly on the uniques / codes after factorize, but that seems like a bad idea and would only help when there aren't a lot of distinct values.

However, we should only need to perform the op when sort is False and dropna is False. I'll confirm that and post a new fix moving this upstream so we can avoid it at least some of the time.

**Pull request merged**